akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -253,9 +253,11 @@ object ByteString {
       else ByteString1(bytes, startIndex, Math.min(n, length))
 
     override def slice(from: Int, until: Int): ByteString = {
-      if (from <= 0 && until >= length) this // we can do < / > since we're Compact
+      if (from <= 0 && length <= until) this // we can do < / > since we're Compact
       else if (until <= from) ByteString1.empty
-      else ByteString1(bytes, startIndex + from, until - from)
+      else if (0 <= from && until <= bytes.length) ByteString1(bytes, startIndex + from, until - from)
+      else if (from < 0) ByteString1(bytes, startIndex, until)
+      else /*(until > bytes.length)*/ ByteString1(bytes, startIndex + from, bytes.length - from)
     }
```

Comment:
I just finished a round of benchmarking, we're OK to go with just `drop().take()` for now and if @monkey-mas has the time to optimise it that would be very nice. Overall HTTP impact is not such that we need to worry about it - difference is within error limits pretty much